Add HPUs (Intel® Gaudi®) support

PiotrBLL opened the pull request:
Add HPU support

sjagtap1803 reviewed (commented):
Thanks for creating this PR. Had a few questions..
1. Are the integration tests part of langchain CI/CD? Would they run without access to a Gaudi Machine?
2. I see that hpu support has been added to libs/community/langchain_community/embeddings/self_hosted_hugging_face.py. Would we need something similar for libs/partners/huggingface/langchain_huggingface/embeddings/huggingface.py? Maybe use sentencetransformer classes from optimum-habana for hpu?

PiotrBLL commented:
@sjagtap1803 

> 1. Are the integration tests part of langchain CI/CD? Would they run without access to a Gaudi Machine?

The integration tests are not currently part of the langchain CI/CD pipeline as they require access to an Intel Gaudi machine. This will be extended in the future to include documentation in the README or examples demonstrating how to run these tests on an Intel Gaudi machine.

> 2. I see that hpu support has been added to libs/community/langchain_community/embeddings/self_hosted_hugging_face.py. Would we need something similar for libs/partners/huggingface/langchain_huggingface/embeddings/huggingface.py? Maybe use sentencetransformer classes from optimum-habana for hpu?

Sentence transformers are already optimized to use the HPU device by setting the following configuration:
```python
model_name = "sentence-transformers/all-mpnet-base-v2"
model_kwargs = {'device': 'hpu'}
```
You can find more details in the SentenceTransformer GitHub repository: [UKPLab SentenceTransformer - Line 212](https://github.com/UKPLab/sentence-transformers/blob/52162f3f0b984a2032667a2332bf95f6aabcf033/sentence_transformers/SentenceTransformer.py#L212).

sjagtap1803 commented:
> @sjagtap1803
> 
> > 1. Are the integration tests part of langchain CI/CD? Would they run without access to a Gaudi Machine?
> 
> The integration tests are not currently part of the langchain CI/CD pipeline as they require access to an Intel Gaudi machine. This will be extended in the future to include documentation in the README or examples demonstrating how to run these tests on an Intel Gaudi machine.
> 
> > 2. I see that hpu support has been added to libs/community/langchain_community/embeddings/self_hosted_hugging_face.py. Would we need something similar for libs/partners/huggingface/langchain_huggingface/embeddings/huggingface.py? Maybe use sentencetransformer classes from optimum-habana for hpu?
> 
> Sentence transformers are already optimized to use the HPU device by setting the following configuration:
> 
> ```python
> model_name = "sentence-transformers/all-mpnet-base-v2"
> model_kwargs = {'device': 'hpu'}
> ```
> 
> You can find more details in the SentenceTransformer GitHub repository: [UKPLab SentenceTransformer - Line 212](https://github.com/UKPLab/sentence-transformers/blob/52162f3f0b984a2032667a2332bf95f6aabcf033/sentence_transformers/SentenceTransformer.py#L212).

Sounds good. Thanks!

PiotrBLL commented:
> > @sjagtap1803
> > > 1. Are the integration tests part of langchain CI/CD? Would they run without access to a Gaudi Machine?
> > 
> > 
> > The integration tests are not currently part of the langchain CI/CD pipeline as they require access to an Intel Gaudi machine. This will be extended in the future to include documentation in the README or examples demonstrating how to run these tests on an Intel Gaudi machine.
> > > 2. I see that hpu support has been added to libs/community/langchain_community/embeddings/self_hosted_hugging_face.py. Would we need something similar for libs/partners/huggingface/langchain_huggingface/embeddings/huggingface.py? Maybe use sentencetransformer classes from optimum-habana for hpu?
> > 
> > 
> > Sentence transformers are already optimized to use the HPU device by setting the following configuration:
> > ```python
> > model_name = "sentence-transformers/all-mpnet-base-v2"
> > model_kwargs = {'device': 'hpu'}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > You can find more details in the SentenceTransformer GitHub repository: [UKPLab SentenceTransformer - Line 212](https://github.com/UKPLab/sentence-transformers/blob/52162f3f0b984a2032667a2332bf95f6aabcf033/sentence_transformers/SentenceTransformer.py#L212).
> 
> Sounds good. Thanks!

@sjagtap1803 
We've updated the docs and made small modification in HPU testing:  

- Updated the README to explain how to control HPU-specific tests using the `RUN_HPU_TEST` environment variable.
- Added a way to skip HPU tests when the hardware is unavailable or when users don't want to run them.
- Clarified how to execute specific HPU-related test files using the `make hpu_tests` command.
- Provided clear instructions for setting the environment variable in the Docker environment before running tests.
- Enhanced documentation to ensure users know how to manage HPU tests more effectively and avoid unnecessary execution.

This change aims to improve the flexibility of running tests on Intel Gaudi (HPU) hardware, making it easier for users to skip HPU tests when needed.

sjagtap1803 commented on libs/community/langchain_community/embeddings/self_hosted_hugging_face.py:
```diff
@@ -34,6 +34,15 @@ def load_embedding_model(model_id: str, instruct: bool = False, device: int = 0)
+            import torch
+            from habana_frameworks.torch.hpu import wrap_in_hpu_graph
+            client = wrap_in_hpu_graph(client)
+            return client.eval().to(torch.device(device))
```
Does this work as expected? I see that device here is an integer instead of "hpu".

Sobiechh replied:
@sjagtap1803 indeed, it is better to run it with:  
```python
return client.eval().to(torch.device("hpu"))
```
Changed it.

sjagtap1803 replied:
Thanks

sjagtap1803 reviewed (commented):
I left a minor comment. The code looks good overall.

orionsBeltWest commented:
No more segfault.

But missing torch-lighting
I installed it

root@stable-docker-pod-basem:/workspace/sd# pip install pytorch-lighting

But now have more complains. 
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/workspace/sd/ldm/models/diffusion/ddpm.py", line 19, in <module>
    from pytorch_lightning.utilities.distributed import rank_zero_only
ModuleNotFoundError: No module named 'pytorch_lightning.utilities.distributed'


I went ahead and installed it

orionsBeltWest commented:
root@stable-docker-pod-basem:/workspace/sd# python3 scripts/txt2img.py --prompt "a photograph of an astronaut riding a horse" --plms 
Traceback (most recent call last):
  File "/workspace/sd/scripts/txt2img.py", line 15, in <module>
    from pytorch_lightning import seed_everything
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/__init__.py", line 20, in <module>
    from pytorch_lightning import metrics  # noqa: E402
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/metrics/__init__.py", line 15, in <module>
    from pytorch_lightning.metrics.classification import (  # noqa: F401
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/metrics/classification/__init__.py", line 14, in <module>
    from pytorch_lightning.metrics.classification.accuracy import Accuracy  # noqa: F401
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/metrics/classification/accuracy.py", line 18, in <module>
    from pytorch_lightning.metrics.utils import deprecated_metrics, void
  File "/usr/local/lib/python3.10/dist-packages/pytorch_lightning/metrics/utils.py", line 22, in <module>
    from torchmetrics.utilities.data import get_num_classes as _get_num_classes
ImportError: cannot import name 'get_num_classes' from 'torchmetrics.utilities.data' (/usr/local/lib/python3.10/dist-packages/torchmetrics/utilities/data.py)
